### DISCUSSION TITLE
Can this use Lancache

### DISCUSSION THREAD

**Discussion by nutterthanos** [ValvePython/steamctl#42] [Q&A] [answered]
I was curious how you could use this program with lancache.
I can't get to use it tho

**Comment by rossengeorgiev** [comment, marked as answer]
Short answer: yes

As far as I understand lancache uses DNS to redirect the requests though itself and cache the data. `steamctl` uses the default resolver, so if you set to point at lancache, then downloads via `steamctl` will get cached, as long as they are not HTTPS.

Alternatively, simply forcing lancache as content server by modifying the local content server list.

Edit `~/.local/share/steamctl/cs_servers.json` to:
```
{
    "cell_id": 1337,
    "servers": [
        {
            "cell_id": 1337,
            "host": "lancache.steamcontent.com",
            "https": false,
            "load": 0,
            "port": 80,
            "type": "SteamCache",
            "vhost": "lancache.steamcontent.com",
            "weighted_load": 0.0
        }
    ],
    "timestamp": 2147483000
}
```

For best download performance, just keep the top 4 servers, which would be geographically close, and remove the rest especially https ones. Setting the timestamp way into the future will keep `steamctl` from updating the file.

**Reply by nutterthanos** [reply]
but they are not being cached
EDIT: well it does resolve dns requests through my lancache but for whatever reason the nginx cache server running is not picking anything up in the access.log file
maybe it only does it for certain process names?
or maybe steamctl does not try to download only over http?
Second Edit: i noticed this in wireshark when using steamctl
![image](https://user-images.githubusercontent.com/50470661/134786966-a9acb249-039c-4e6d-a9c7-e0979a170b12.png)

**Reply by rossengeorgiev** [reply]
Try setting the nameserver on the host you are running `steamctl` or depotdownloader on to the lancache dns server. Both applications act the same as the Steam app. I don't see why it wouldn't work

**Reply by nutterthanos** [reply]
yeah my dns server on my host is set to use my lancache dns server

**Reply by nutterthanos** [reply]
> Try setting the nameserver on the host you are running `steamctl` or depotdownloader on to the lancache dns server. Both applications act the same as the Steam app. I don't see why it wouldn't work

i asked this in lancache.net discord server
![image](https://user-images.githubusercontent.com/50470661/135183314-7e4b793d-d7e0-4a35-a5a2-04d353e3fdd9.png)
this same logic applies to any client accessing steam

**Reply by rossengeorgiev** [reply]
Perhaps I misunderstand how landcache works. I assumed that when using the landcache DNS server for DNS resolution it responds for any query for Steam CDN server with lancache instead of the original response.  Essentially redirecting CDN request to lancache. This would be transparent to the steam client or anything else

**Reply by nutterthanos** [reply]
> Perhaps I misunderstand how landcache works. I assumed that when using the landcache DNS server for DNS resolution it responds for any query for Steam CDN server with lancache instead of the original response. Essentially redirecting CDN request to lancache. This would be transparent to the steam client or anything else

yeah maybe you did misunderstood how it works
but how could we get to to work then?
Edit: https://github.com/SteamRE/SteamKit/issues/839
maybe your code is forcing https?

**Reply by nutterthanos** [reply]
![image](https://user-images.githubusercontent.com/50470661/135391123-d14ea1be-582c-466f-8ceb-1c223e04bf4e.png)
this looks like code in https://github.com/ValvePython/steam/blob/master/steam/client/cdn.py which looks like it is forcing https?

**Reply by rossengeorgiev** [reply]
I had a little play with lancache, and it works just as I expected. If you are using lancache dns as your resolver, steamctl will hit through lancache for the most part as very few content servers are mandatory https. However, you can force `steamctl` to only use lancache by modifying the content server list stored locally.

Edit `~/.local/share/steamctl/cs_servers.json` like so:
```
{
    "cell_id": 1337,
    "servers": [
        {
            "cell_id": 1337,
            "host": "lancache.steamcontent.com",
            "https": false,
            "load": 0,
            "port": 80,
            "type": "SteamCache",
            "vhost": "lancache.steamcontent.com",
            "weighted_load": 0.0
        }
    ],
    "timestamp": 2147483000
}
```

**Reply by nutterthanos** [reply]
> I had a little play with lancache, and it works just as I expected. If you are using lancache dns as your resolver, steamctl will hit through lancache for the most part as very few content servers are mandatory https. However, you can force `steamctl` to only use lancache by modifying the content server list stored locally.
> 
> Edit `~/.local/share/steamctl/cs_servers.json` like so:
> 
> ```
> {
>     "cell_id": 1337,
>     "servers": [
>         {
>             "cell_id": 1337,
>             "host": "lancache.steamcontent.com",
>             "https": false,
>             "load": 0,
>             "port": 80,
>             "type": "SteamCache",
>             "vhost": "lancache.steamcontent.com",
>             "weighted_load": 0.0
>         }
>     ],
>     "timestamp": 2147483000
> }
> ```
but if you only put above in cs_servers.json would it only grab files from the lancache or would it still download like files from upstream?

**Reply by rossengeorgiev** [reply]
It would only downloading files from lancache. Landcache downloads from upstream if the files not in available in the cache

**Comment by nutterthanos** [comment]
thanks this worked

**Reply by nutterthanos** [reply]
only thing i seem to notice is its downloading slow when a file is a "MISS"

**Reply by rossengeorgiev** [reply]
`lancache.steamcontent.com` likely resolves to not ideal content servers. I suggest remove the `cs_servers.json`, and run `steamctl` to get new one. That list of servers will be close to you. Edit the list, and leave top 4 servers. Then edit `timestamp` to `2147483000` in order to prevent `steamctl` from updating file any time soon.

You can still experience low download speeds if you have an extremely good connection as `steamctl` is single threaded. Once the process hits 100% cpu use that is about as good as download speed you are going get until it's fixed

**Reply by nutterthanos** [reply]
i only.seem to get this in my cs_servers.json file
{
    "cell_id": 54,
    "servers": [
        {
            "cell_id": 0,
            "host": "steampipe.akamaized.net",
            "https": true,
            "load": 0,
            "port": 443,
            "type": "CDN",
            "vhost": "steampipe.akamaized.net",
            "weighted_load": 130
        }
    ],
    "timestamp": 1633140413
}

**Reply by nutterthanos** [reply]
@rossengeorgiev nvm finally manged to get it to work by doing like this 
![image](https://user-images.githubusercontent.com/50470661/154632915-ef92613f-585e-426f-a941-b2b491688c99.png)
